_drafts/partially-ordering-time.md

```diff
@@ -0,0 +1,253 @@
+# Time Is Weird
+## or: why do distributed systems and weak memory models look so similar, anyway
+
+(_Epistemic status: I'm just learning about both of these fields, and these are simply my
+observations from studying them and comparing them. If you work in either field and
+something looks wrong, it probably is; let me know!_)
+
+    There's only one hard problem in computer science: recognising that cache invalidation errors
+    are misnamed. They're just off-by-one errors in the time domain.
+        —Unknown
+
+Time is weird.
+
+Time is weird, not just because we insist on trying to [measure it](#url) and [standardise
+it](#url), though that field is of course amazingly horrifying. Time is weird, because
+when we try to think about what it actually _implies_, that a second has passed since one
+second ago.
+
+The best we've been able to come up with about the arrow of time is that it in some way
+tracks things that “happened” _before_ other things that “happened” _after_ them, in the
+sense, that, uh, the later events have access to information? and entropy? from the
+earlier events? Which means, um, something something flow of causality?
+
+Of course that breaks down, at distances comparable to light speed. When you can no longer
+assume that events are instantaneously knowable, time behaves _seriously_ anomalously. Can
+an alien reading about memory models outside our light cone meaningfully have done so
+_before_ you did?
+
+But we ignore that, and hope that in the vast majority of cases we can just assume time
+just works how we kinda normally want it to work. There's no reason to make life so much
+more complicated for ourselves, right?
+
+…if there's one thing I've understood about computer folk, it's that we _never_ pass up a
+chance to make life more complicated for ourselves.
+
+### Time, But Only Sometimes
+
+Time is weird.
+
+Time is weird because there are so many cases where (processors / compilers / “the
+network”, pick all) look at two events that “happened” in one order and decide, nah, those
+don't need to “happen” in that order.
+
+Your hardware will have a memory ordering model; this captures what license your CPU takes
+when executing assembly as to which instructions happen before which other ones. (Or is
+supposed to, anyway; in practice many hardware memory models are
+[underspecified](#urls[1-2]).)
+
+Your compiler will have a memory model; this captures how your language chooses to expose
+the hardware memory model to you, as well as what license your compiler takes in
+reordering instructions _while generating assembly_ across multiple threads. This is what
+you code to when writing lock-free systems code.
+
+In practice, there's a single weak memory model that matters to most people—the C++11
```

Comment:
definitely needs a short definition/discussion of weak vs strong memory models.

Reply:
and maybe what the motivation is for having strong and weak memory models? what's the tl;dr version of what they solve, enable, or prevent?